src/Datastore/DatastoreSessionHandler.php

```diff
@@ -0,0 +1,277 @@
+ * $handler = new DatastoreSessionHandler(new DatastoreClient());
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * # Then read and write the $_SESSION array.
+ *
+ * function handle_session_error($errNo, $errStr, $errFile, $errLine) {
+ *     # We throw an exception here, but you can do whatever you need.
+ *     throw new Exception("$errStr in $errFile on line $errLine", $errNo);
+ * }
+ * set_error_handler('handle_session_error', E_USER_WARNING);
+ * # If `write` fails for any reason, an exception will be thrown.
+ * session_write_close();
+ * restore_error_handler();
+ * # You can still read the $_SESSION array after closing the session.
+ * ```
+ */
+class DatastoreSessionHandler implements SessionHandlerInterface
+{
+    const DEFAULT_GC_LIMIT = 0;
+    const DEFAULT_KIND = 'PHPSESSID';
+    const DEFAULT_NAMESPACE_ID = 'sessions';
+    /*
+     * {@see https://cloud.google.com/datastore/docs/reference/rpc/google.datastore.v1#google.datastore.v1.PartitionId}
+     */
+    const NAMESPACE_ALLOWED_PATTERN = '/^[A-Za-z\d\.\-_]{0,100}$/';
+    const NAMESPACE_RESERVED_PATTERN = '/^__.*__$/';
+
+    /* @var int */
+    private $gcLimit;
+
+    /* @var DatastoreClient */
+    private $datastore;
+
+    /* @var string */
+    private $kind;
+
+    /* @var string */
+    private $namespaceId;
+
+    /* @var Transaction */
+    private $transaction;
+
+    /**
+     * @param DatastoreClient $datastore Datastore client.
+     * @param int [optional] $gcLimit A number of entities to delete in the
+     *        garbage collection.  Defaults to 0 which means it does nothing.
+     *        The value larger than 1000 will be cut down to 1000.
+     */
+    public function __construct(
+        DatastoreClient $datastore,
+        $gcLimit = self::DEFAULT_GC_LIMIT
+    ) {
+        $this->datastore = $datastore;
+        // Cut down to 1000
+        $this->gcLimit = min($gcLimit, 1000);
+        $this->kind = self::DEFAULT_KIND;
+        $this->namespaceId = self::DEFAULT_NAMESPACE_ID;
+    }
+
+    /**
+     * @param string $savePath The value of `session.save_path` setting will be
+     *        used here. It will use this value as the Datastore namespaceId.
+     * @param string $sessionName The value of `session.name` setting will be
+     *        used here. It will use this value as the Datastore kind.
+     * @return bool
+     */
+    public function open($savePath, $sessionName)
+    {
+        $this->kind = $sessionName;
+        if (preg_match(self::NAMESPACE_ALLOWED_PATTERN, $savePath) !== 1 ||
+            preg_match(self::NAMESPACE_RESERVED_PATTERN, $savePath) === 1) {
+            throw new InvalidArgumentException(
+                sprintf('The given save_path "%s" not allowed', $savePath)
+            );
+        }
+        $this->namespaceId = $savePath;
+        $this->transaction = $this->datastore->transaction();
+        return true;
+    }
+
+    public function close()
```

Comment:
I don't think we implement support for `@inheritDoc`. It might work though.

You can exclude it from being listed in the list of methods with `@access private` if you feel that it doesn't need to be shown on the doc page at all.

```
/**
 * @access private
 */
```